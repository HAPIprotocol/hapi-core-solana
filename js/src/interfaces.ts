import {
  Commitment,
  Connection,
  ConnectionConfig,
  Keypair,
  PublicKey,
} from "@solana/web3.js";

export interface HapiClientConfig {
  endpoint: string | Connection;
  commitment?: Commitment | ConnectionConfig;
  communityName?: string;
  programId?: string;
}

export interface HapiViewResponse<T> {
  account: PublicKey;
  data: T;
}

export interface HapiActionResponse<T> {
  txHash: string;
  account: PublicKey;
  data: T;
}

export interface HapiActionResponseWithMeta<T, U> {
  txHash: string;
  account: PublicKey;
  data: T;
  meta: U;
}

export interface HapiClientAuthorityConfig extends HapiClientConfig {
  payer: Keypair | PublicKey;
}

export interface HapiClientReporterConfig extends HapiClientConfig {
  payer: Keypair | PublicKey;
}
