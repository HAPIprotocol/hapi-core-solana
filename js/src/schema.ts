import { SOLANA_SCHEMA } from "@solana/web3.js";

export const PROGRAM_SCHEMA = SOLANA_SCHEMA;
