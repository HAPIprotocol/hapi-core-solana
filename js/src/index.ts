export * from "./state";
export * from "./constants";
export { u64 } from "./utils";
export * from "./reader-client";
export * from "./reporter-client";
export * from "./authority-client";
export * from "./instructions/instructions";
