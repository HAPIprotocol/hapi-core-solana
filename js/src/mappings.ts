import {
  HapiAccountType,
  HapiInstruction,
  ReporterType,
  Category,
  CaseStatus,
} from "./state/enums";

const HAPI_INSTRUCTION_NAMES = {
  [HapiInstruction.CreateCommunity]: "CreateCommunity",
  [HapiInstruction.UpdateCommunity]: "UpdateCommunity",
  [HapiInstruction.CreateNetwork]: "CreateNetwork",
  [HapiInstruction.UpdateNetwork]: "UpdateNetwork",
  [HapiInstruction.CreateReporter]: "CreateReporter",
  [HapiInstruction.UpdateReporter]: "UpdateReporter",
  [HapiInstruction.CreateCase]: "CreateCase",
  [HapiInstruction.UpdateCase]: "UpdateCase",
  [HapiInstruction.CreateAddress]: "CreateAddress",
  [HapiInstruction.UpdateAddress]: "UpdateAddress",
};
export function getHapiInstructionName(instruction: HapiInstruction): string {
  return HAPI_INSTRUCTION_NAMES[instruction] || "<Unknown>";
}

const HAPI_ACCOUNT_TYPE_NAMES = {
  [HapiAccountType.Uninitialized]: "Uninitialized",
  [HapiAccountType.Community]: "Community",
  [HapiAccountType.Network]: "Network",
  [HapiAccountType.Reporter]: "Reporter",
  [HapiAccountType.Case]: "Case",
  [HapiAccountType.Address]: "Address",
};
export function getHapiAccountTypeName(accountType: HapiAccountType): string {
  return HAPI_ACCOUNT_TYPE_NAMES[accountType] || "<Unknown>";
}

const REPORTER_TYPE_NAMES = {
  [ReporterType.Inactive]: "Inactive",
  [ReporterType.Tracer]: "Tracer",
  [ReporterType.Full]: "Full",
  [ReporterType.Authority]: "Authority",
};
export function getReporterTypeName(reporterType: ReporterType): string {
  return REPORTER_TYPE_NAMES[reporterType] || "<Unknown>";
}

const CATEGORY_NAMES = {
  [Category.Safe]: "Safe",
  [Category.WalletService]: "WalletService",
  [Category.MerchantService]: "MerchantService",
  [Category.MiningPool]: "MiningPool",
  [Category.LowRiskExchange]: "LowRiskExchange",
  [Category.MediumRiskExchange]: "MediumRiskExchange",
  [Category.DeFi]: "DeFi",
  [Category.OTCBroker]: "OTCBroker",
  [Category.ATM]: "ATM",
  [Category.Gambling]: "Gambling",
  [Category.IllicitOrganization]: "IllicitOrganization",
  [Category.Mixer]: "Mixer",
  [Category.DarknetService]: "DarknetService",
  [Category.Scam]: "Scam",
  [Category.Ransomware]: "Ransomware",
  [Category.Theft]: "Theft",
  [Category.Counterfeit]: "Counterfeit",
  [Category.TerroristFinancing]: "TerroristFinancing",
  [Category.Sanctions]: "Sanctions",
  [Category.ChildAbuse]: "ChildAbuse",
};
export function getCategoryName(category: Category): string {
  return CATEGORY_NAMES[category] || "<Unknown>";
}

const CASE_STATUS_NAMES = {
  [CaseStatus.Open]: "Open",
  [CaseStatus.Closed]: "Closed",
};
export function getCaseStatusName(caseStatus: CaseStatus): string {
  return CASE_STATUS_NAMES[caseStatus] || "<Unknown>";
}
