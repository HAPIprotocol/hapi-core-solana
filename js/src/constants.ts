import { PublicKey } from "@solana/web3.js";

export const HAPI_PROGRAM_ID = new PublicKey(
  "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7"
);
