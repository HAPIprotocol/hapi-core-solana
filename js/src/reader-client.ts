import {
  Commitment,
  Connection,
  ConnectionConfig,
  PublicKey,
} from "@solana/web3.js";

import { base58ToPublicKey, u64 } from "./utils";
import { Address, Case, Community, Network, Reporter } from "./state";
import { HAPI_PROGRAM_ID } from "./constants";

export interface HapiClientConfig {
  endpoint: string;
  commintment?: Commitment | ConnectionConfig;
  communityName?: string;
  programId?: string;
}

export interface HapiViewResponse<T> {
  account: PublicKey;
  data: T;
}

export interface HapiActionResponse<T> {
  account: PublicKey;
  txHash: string;
  data: T;
}

/** HAPI client to read entity data from Solana */
export class ReaderClient {
  protected connection: Connection;
  protected communityName?: string;
  protected programId: PublicKey;

  constructor(config: HapiClientConfig) {
    this.connection = new Connection(config.endpoint, config.commintment);
    this.communityName = config.communityName;
    this.programId = new PublicKey(this.programId || HAPI_PROGRAM_ID);
  }

  /**
   * Sets community name for context
   * @param communityName Community name
   * @returns Self
   */
  switchCommunity(communityName: string): ReaderClient {
    this.communityName = communityName;
    return this;
  }

  /**
   * Fetch community info from blockchain
   * @param communityName (Optional) The name of the community to fetch (defaults to context)
   * @returns Community info
   **/
  async getCommunity(
    communityName?: string
  ): Promise<HapiViewResponse<Community>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Community.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName
    );

    return state;
  }

  /**
   * Fetch network info from blockchain
   * @param networkName The name of the network to fetch
   * @param communityName (Optional) The name of the community to fetch network from (defaults to context)
   * @returns Network info
   **/
  async getNetwork(
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Network>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const state = await Network.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName
    );

    return state;
  }

  /**
   * Fetch reporter info from blockchain
   * @param reporterPubkey Public key of the reporter to fetch
   * @param communityName (Optional) The name of the community to fetch reporter from (defaults to context)
   * @returns Reporter info
   **/
  async getReporter(
    reporterPubkey: string,
    communityName?: string
  ): Promise<HapiViewResponse<Reporter>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Reporter.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      new PublicKey(reporterPubkey)
    );

    return state;
  }

  /**
   * Fetch case info from blockchain
   * @param caseId ID of the case to fetch
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Case info
   **/
  async getCase(
    caseId: u64,
    communityName?: string
  ): Promise<HapiViewResponse<Case>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Case.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      caseId
    );

    return state;
  }

  /**
   * Fetch address info from blockchain
   * @param address The address to fetch info for (string for Solana addresses, Buffer for others)
   * @param networkName The name of the network to which address belongs to
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Address info
   **/
  async getAddress(
    address: string | Buffer,
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Address>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const convertedAddress =
      address instanceof Buffer
        ? new PublicKey(address)
        : base58ToPublicKey(address);

    const state = await Address.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName,
      convertedAddress
    );

    return state;
  }
}
