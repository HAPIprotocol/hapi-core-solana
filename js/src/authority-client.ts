import {
  Transaction,
  TransactionInstruction,
  PublicKey,
  Keypair,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
  sendAndConfirmTransaction,
} from "@solana/web3.js";

import { Community, HapiAccountType } from "./state";
import { HapiActionResponse, ReaderClient } from "./reader-client";
import { HAPI_PROGRAM_ID } from "constants";

export class AuthorityClient extends ReaderClient {
  async createCommunity(
    communityName: string
  ): Promise<HapiActionResponse<Community>> {
    let existingAccount: PublicKey;
    try {
      const existing = await this.getCommunity(communityName);
      existingAccount = existing.account;
    } catch {
      // NOP
    }

    if (existingAccount) {
      throw new Error(
        `Community already exists: ${communityName} (${existingAccount})`
      );
    }

    const authority = new Keypair();

    const [communityAddress] = await Community.getAddress(communityName);

    const transaction = new Transaction();

    {
      SystemProgram.createAccount({
        space: Community.size,
        programId: 
      })

      // TODO: zzzzzzzzzzzzzzz....

      const instruction = new TransactionInstruction({
        keys: [
          { pubkey: authority.publicKey, isSigner: true, isWritable: true },
          { pubkey: communityAddress, isSigner: false, isWritable: true },
          { pubkey: authority.publicKey, isSigner: false, isWritable: false, }
        ],
        programId: HAPI_PROGRAM_ID,
        data: Buffer.alloc(0),
      });
    }

    {
      const community = new Community({
        accountType: HapiAccountType.Community,
        name: communityName,
        authority: authority.publicKey,
      });

      const instruction = new TransactionInstruction({
        keys: [
          { pubkey: authority.publicKey, isSigner: true, isWritable: false },
          { pubkey: communityAddress, isSigner: false, isWritable: true },
          {
            pubkey: SystemProgram.programId,
            isSigner: false,
            isWritable: false,
          },
          { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
        ],
        programId: HAPI_PROGRAM_ID,
        data: Buffer.from(community.serialize()),
      });

      transaction.add(instruction);
    }
    const signature = await sendAndConfirmTransaction(
      this.connection,
      transaction,
      [authority],
      { commitment: "confirmed" }
    );

    // TODO: create an instruction
    return {
      publicKey: communityAddress,
      txHash: signature,
    };
  }

  async createNetwork(): Promise<void> {
    // TODO: create an instruction
  }

  async createReporter(): Promise<void> {
    // TODO: create an instruction
  }

  async updateReporter(): Promise<void> {
    // TODO: create an instruction
  }
}
