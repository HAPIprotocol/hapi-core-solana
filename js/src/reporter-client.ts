import { PublicKey } from "@solana/web3.js";

import { HapiActionResponse, ReaderClient } from "./reader-client";
import { Case, Category } from "./state";

/** HAPI client to operate reporter program functions on Solana */
export class ReporterClient extends ReaderClient {
  async createCase(
    name: string,
    categories: Category[],
    communityName?: string
  ): Promise<HapiActionResponse<Case>> {
    name;
    categories;
    communityName;
    // TODO: create an instruction
    return {
      account: new PublicKey("0"),
    };
  }

  async updateCase(): Promise<void> {
    // TODO: create an instruction
  }

  async createAddress(): Promise<void> {
    // TODO: create an instruction
  }

  async updateAddress(): Promise<void> {
    // TODO: create an instruction
  }
}
