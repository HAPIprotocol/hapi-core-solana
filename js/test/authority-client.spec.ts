import { AuthorityClient } from "../src";

describe("AuthorityClient", () => {
  let client: AuthorityClient;

  beforeEach(() => {
    client = new AuthorityClient({
      endpoint: "http://localhost:8899",
    });
  });

  it("should initialize", () => {
    expect(client).toBeDefined();
  });

  describe("createCommunity", () => {
    it("should throw - community already exists", async () => {
      await expect(() =>
        client.createCommunity("hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
    it("should throw - invalid name", async () => {
      await expect(() =>
        client.createCommunity("loooooooooooooooooooooooooooooooooong")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
    it.todo("should create a community - success");
  });

  describe("createNetwork", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - network already exists");
    it.todo("should throw - invalid name");
    it.todo("should create a network - success");
  });

  describe("createReporter", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - reporter already exists");
    it.todo("should throw - invalid type");
    it.todo("should throw - invalid name");
    it.todo("should throw - invalid public key");
    it.todo("should create a reporter - success");
  });

  describe("updateReporter", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - reporter not found");
    it.todo("should throw - invalid type");
    it.todo("should throw - invalid name");
    it.todo("should update a reporter - success");
  });
});
