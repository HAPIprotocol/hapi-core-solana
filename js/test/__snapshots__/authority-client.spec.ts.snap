// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorityClient createCommunity should throw - community already exists 1`] = `"Community already exists: hapi.one (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;
